### DISCUSSION TITLE
GradCAM generates same results for every slice when using 3D DenseNet121

### DISCUSSION THREAD

**Discussion by gyumin1ee** [Project-MONAI/MONAI#7172] [Q&A]
Hello,

The GradCAM I generated has a shape of (1, 1, 256, 256, 50), but it produces the same results for every slice.

I am not very familiar with manipulating tensors so i don`t know where should i fix the code.

How can i get different result per slice in one image? 

Thank you

```
from monai.visualize import GradCAM

# print(model)
device = torch.device('cuda')
cam_extractor = GradCAM(nn_module=model, target_layers="features.denseblock4.denselayer16") 

# features.denseblock4.denselayer16

num_correct = 0.0
metric_count = 0
datacount = 0

model.load_state_dict(torch.load('best_metric_model_classification3d_dict.pth'))
model.eval().to(device)
datacount = 0
for data in tqdm(test_loader, leave=False):
    inputs= data["image"].to(device)#torch.Size([1, 1, 256, 256, 50])
    out = model(inputs) #torch.Size([1, 2])
    cams = cam_extractor(inputs) #torch.Size([1, 1, 256, 256, 50])

    plt.figure(figsize=(45, 30))
    
    slicecount = 0
    
    for image in cams.squeeze(0).squeeze(0).permute(2,1,0):    
        plt.subplot(5, 10, slicecount+1)
        plt.imshow(image.cpu().numpy())
        plt.axis('off')
        slicecount = slicecount + 1

    plt.tight_layout()
    plt.savefig(str(datacount) + '_gcam.png')
    datacount = datacount + 1
```

**Comment by KumoLiu** [comment]
Hi @gyumin1ee, I don't see any obvious errors in your code, maybe you can refer to this [tutorial](https://github.com/Project-MONAI/tutorials/blob/main/modules/interpretability/class_lung_lesion.ipynb) or check the output of the network.
Thanks!

**Reply by gyumin1ee** [reply]
Hello @KumoLiu, thank you for your reply!

I checked the output size as you mentioned, and it was the problem.

The number of slices is much smaller than the image width and height, resulting in an imbalance in the input shape and yielding a thin feature map, I think.

I'm going to edit some layers in the model, such as adjusting the kernel size of convolution layers, but I'm not sure that if this is the correct method to fix this problem.

Do you have any suggestions?

**Reply by KumoLiu** [reply]
Hi @gyumin1ee, yes you can always modify the network. But by checking the output of the network I mean if the network has learned anything or if it's just random. It looks weird to have it the same for every slice. I still strongly recommend you start with this [tutorial](https://github.com/Project-MONAI/tutorials/blob/main/modules/interpretability/class_lung_lesion.ipynb) to see if you can achieve the same results and then replace it with your network.
